### PR TITLE
Wiki side of slim expressions

### PR DESCRIPTION
**Pull request by tcnh**
Looking at the acceptance test page, I'd say this covers all cases where symbols are set on the Slim runner side.

I don't see a solution yet on using expressions on symbols that are only set on the executor side (i.e. maps set by map of maps fixture) and I am starting to doubt there is one, given the strict separation of these two

### CODE REVIEW

File: src/fitnesse/testsystems/slim/tables/SlimTable.java [hunk 1]
```diff
@@ -200,7 +203,28 @@ public void setCustomComparatorRegistry(CustomComparatorRegistry customComparato
...
 
-  /** SlimExpectation base class for row based expectations. */
+  private String getSlimExpressionResult(String variableNameWithDollar) {
+    SlimExpressionEvaluator evaluator = new SlimExpressionEvaluator();
+    String expr = variableNameWithDollar.substring(2, variableNameWithDollar.length() - 1);
+    String variableName = expr.split("\\.")[0];
+
+    if (getSymbol(variableName) != null) {
+      Map<String, MethodExecutionResult> symbols = new HashMap<>();
+      symbols.put(variableName, new MethodExecutionResult(getSymbol(variableName), Object.class));
```
Comment (fhoeben) [inline src/fitnesse/testsystems/slim/tables/SlimTable.java]: Shouldn't all symbols from the test's context be made accessible to the evaluator (the expression might reference more than one variable, and might not start with a variable's name). I mean the expression could be `1.5 * myHashMap.size() + myHashMap2.key2.length()`. 
Or am I misinterpreting something?

File: src/fitnesse/slim/SlimExpressionEvaluator.java [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
+import fitnesse.wikitext.parser.*;
 
 import javax.script.ScriptEngine;
 import javax.script.ScriptEngineManager;
 import javax.script.ScriptException;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
```
Comment (fhoeben) [inline src/fitnesse/slim/SlimExpressionEvaluator.java]: This import I was not expecting....

File: src/fitnesse/slim/SlimExpressionEvaluator.java [hunk 3]
```diff
@@ -34,6 +41,9 @@ public void setContext(Map<String, MethodExecutionResult> variables) {
     for (Map.Entry<String, MethodExecutionResult> entry : variables.entrySet()) {
       String key = entry.getKey();
       Object value = entry.getValue().getObject();
+      if(value.toString().startsWith("!{")) {
```
Comment (fhoeben) [inline src/fitnesse/slim/SlimExpressionEvaluator.java]: Shouldn't the handling for 'wiki side' expressions be separate from the 'slim server side'?
For slim-server-side I don't expect to ever encounter this prefix, correct?

**Review by fhoeben — COMMENTED** [review]
Start is there but....

**Comment by tcnh** [comment]
You're very right :)

I've separated the wiki markup parsing using an extra helper, all symbols are now availablet to the expression evaluator (needed an extra method in SlimTestContext to get all symbols). Did some cleanup on the ugly imports..

I think it looks better now

**Comment by tcnh** [comment]
Hehe it helps to tick all the boxes when committing. I agree using multiple expressions in one cell is not a good idea.
in-cell concatenating isn't working right now. Needs some attention. I'll see if I can fix that tomorrow

**Comment by tcnh** [comment]
I think it's pretty much complete now. What do you think? 
Did I overlook anything, or is it time to create a new PR on the fitnesse repo?

File: src/fitnesse/slim/SlimExpressionEvaluator.java [hunk 4]
```diff
@@ -5,8 +5,7 @@
 import javax.script.ScriptEngine;
 import javax.script.ScriptEngineManager;
 import javax.script.ScriptException;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (fhoeben) [inline src/fitnesse/slim/SlimExpressionEvaluator.java]: The changes in this file do not seem to add much value, correct?

File: src/fitnesse/testsystems/slim/tables/WikiSymbolTranslateUtil.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Helper class to translate wiki markup in symbols to a Slim-readable html String
+ */
+
```
Comment (fhoeben) [inline src/fitnesse/testsystems/slim/tables/WikiSymbolTranslateUtil.java:13]: Why the empty line?

**Review by fhoeben — APPROVED** [review]
Looks quite good. Ideally I would like the Slim expression evaluation (or any Slim symbol's value actually) to be pushed to the Slim server, so there would be no need to messing around with (possibly outdated) String values on the wiki side. Something to the extend (as adding a command to do `|$var=|clone symbol|$var|`, where var might also be an expression between backticks ``)
I fear that might require changes in the Slim protocol however, because I'm not sure all Slim servers support `clone symbol`. 

Do you have any clue how hard it would be to add an extra slim command (e.g. `|$var=|clone symbol|$var|`) when the server is known to support it?

**Comment by tcnh** [comment]
I think the problem is that clone symbol also lives on the slim server side (helper implementing statementExecutionerInterface). And if I read it right, the wiki-table side evaluation happens before anything on the server side is executed.

Life would be so much easier if there was only one administration of slim symbols, but I'm afraid that's not going to happen anytime soon :)

**Pull request closed without merging**